iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/util/DeviceRegistryTokenAuthProvider.java

```diff
@@ -36,17 +41,55 @@
 
     private final NamespacedKubernetesClient client;
     private final AuthApi authApi;
+    private final EmbeddedCacheManager cacheManager = new DefaultCacheManager();
+    private Cache<String, TokenReview> tokens;
+    private Cache<String, Boolean> authorizations;
 
-    public DeviceRegistryTokenAuthProvider() {
+    public DeviceRegistryTokenAuthProvider(Duration tokenExpiration) {
+        log.info("Using token cache expiration of {}", tokenExpiration);
         this.client = new DefaultKubernetesClient();
         this.authApi = new KubeAuthApi(this.client, this.client.getConfiguration().getOauthToken());
+        ConfigurationBuilder config = new ConfigurationBuilder();
+        config.expiration().lifespan(tokenExpiration.toSeconds(), TimeUnit.SECONDS);
+        cacheManager.defineConfiguration("tokens", config.build());
+        cacheManager.defineConfiguration("subjects", config.build());
+        tokens = cacheManager.getCache("tokens");
+        authorizations = cacheManager.getCache("subjects");
+
     }
 
     @Override
     public void authenticate(JsonObject authInfo, Handler<AsyncResult<User>> resultHandler) {
 
         final String token = authInfo.getString(TOKEN);
+        tokens.computeIfAbsentAsync(token, review -> authApi.performTokenReview(token))
```

Comment:
The `performTokenReview` might block, I think we need an async version, or put it on a worker thread pool.

Reply:
Same as above